### PR TITLE
✨ Add `CloudAnnData`

### PR DESCRIPTION
**Pull request by falexwolf**

### CODE REVIEW

**Comment by Koncopd** [comment]
Thx, left some comments.

**Comment by Koncopd** [comment]
As for the overall design, you are absolutely right, having something like with with direct access to metadata and with the ability to have streaming subsets of X, layers etc it should be much more convenient that current subset / stream functionality.

File: lndb_storage/object/_subset_anndata.py [hunk 1]
```diff
@@ -96,3 +97,37 @@ def _subset_anndata_file(
...
+            df = get_obs_var_storage(storage, which)
+        return df
+
+    @property
+    def obs(self) -> pd.DataFrame:
+        return self._get_obs_var(which="obs")
+
+    @property
+    def var(self) -> pd.DataFrame:
+        return self._get_obs_var(which="var")
```
[the remark this answers is not in the document]
Comment (falexwolf) [reply lndb_storage/object/_subset_anndata.py:137]: Done!

File: lndb_storage/object/_subset_anndata.py [hunk 2]
```diff
@@ -96,3 +97,37 @@ def _subset_anndata_file(
...
+
+
+class CloudAnnData:
+    def __init__(self, file: File):
+        self._file = file
+
+    def _get_obs_var(self, which=Literal["obs", "var"]) -> pd.DataFrame:
+        file_path = filepath_from_file(self._file)
+        fs, file_path_str = _infer_filesystem(file_path)
+        if self._file.suffix == ".h5ad":
```
[the remark this answers is not in the document]
Comment (falexwolf) [reply lndb_storage/object/_subset_anndata.py:121]: Agreed! I just couldn't see how to do it!

**Comment by falexwolf** [comment]
> As for the overall design, you are absolutely right, having something like with with direct access to metadata and with the ability to have streaming subsets of X, layers etc it should be much more convenient that current subset / stream functionality.

Cool! Then, let's work in this direction!

But before doing so let's be wise and strategical how we balance across AnnData, MuData, SOMA/TileDB, plain zarr, zarr for images, etc.

I'd put a focus on scRNA-seq, of course, but maybe we don't need to over-engineer if we can rather build a layer for this whole class of infra.

File: lndb_storage/object/_subset_anndata.py [hunk 3]
```diff
@@ -96,3 +98,40 @@ def _subset_anndata_file(
...
     return adata
+
+
+def get_obs_var_storage(
+    storage: Union[zarr.Group, h5py.File], which=Literal["obs", "var"]
+) -> pd.DataFrame:
+    with storage as access:
+        array = read_elem(access[which])
+    if "index" in array.dtype.fields:
+        return pd.DataFrame.from_records(array, index="index")
```
Comment (Koncopd) [inline lndb_storage/object/_subset_anndata.py:109]: Strange, it should be DataFrame. Probably it is array only for older adatas.

Comment (falexwolf) [reply lndb_storage/object/_subset_anndata.py:109]: It returned an array for my example. Hm. Should I add a test?

Comment (falexwolf) [reply lndb_storage/object/_subset_anndata.py:109]: You're right:

Comment (falexwolf) [reply lndb_storage/object/_subset_anndata.py:109]: Gives me an error here.

File: lndb_storage/object/_subset_anndata.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Koncopd) [reply lndb_storage/object/_subset_anndata.py:124]: I can rework when you merge.

File: lndb_storage/object/_subset_anndata.py (hunk 3 above)
Comment (Koncopd) [reply lndb_storage/object/_subset_anndata.py]: It seems to be array only for old adatas.

**Pull request merged**